### DISCUSSION TITLE
Template language support

### DISCUSSION THREAD

**Discussion by r7kamura** [rubocop/rubocop#10768] [Ideas]
It would be nice if RuboCop would support template engines that embeds Ruby (e.g. ERB, Haml, Slim, etc.), or provide an API for 3rd party plugins.

I'm developing the following tools:

- https://github.com/r7kamura/erbcop
- https://github.com/r7kamura/slimcop
- https://github.com/r7kamura/hamlcop
- https://github.com/r7kamura/templatecop

but it would be hard to implement the same CLI functionality as `rubocop` for each template engine, and doing so would lead to unsupported options like  `--auto-gen-config` and `--server`, etc.

FYI:

How these tools are implemented now is:

1. extract the Ruby code from template
2. process it into an instance of `RuboCop::ProcessedSource`
3. [pass it to `RuboCop::Cop::Team#investigate`](https://github.com/r7kamura/templatecop/blob/beac06b4b2ae37713ffed94bd724d91b94ea7645/lib/templatecop/ruby_offense_collector.rb#L24)
4. [fix offense location based on the location of the embedded Ruby code](https://github.com/r7kamura/templatecop/blob/beac06b4b2ae37713ffed94bd724d91b94ea7645/lib/templatecop/offense.rb#L51-L52)

**Comment by r7kamura** [comment]
For example, when creating rubocop-erb, the following functions should be provided by rubocop-erb:

1. what file patterns to target: *.erb
2. what default config to provide in this template language: e.g. disable Style/FrozenStringLiteralComment, ... 
3. the logic to extract multiple Ruby codes from one ERB file with location information

I think that the rubocop side can be well abstracted if a Ruby file is treated as a kind of a template of Ruby itself and implemented through something like rubocop-ruby 🤔

**Comment by bbatsov** [comment]
Sorry for the slow response here - I'm open to the idea, but I also want to avoid having too many runtime dependencies in RuboCop itself, as hopefully not everyone's using Ruby for Web development. 

> It would be nice if RuboCop would support template engines that embeds Ruby (e.g. ERB, Haml, Slim, etc.), or provide an API for 3rd party plugins.

It seems to me that providing the a proper API for the 3rd party packages would be the simpler solution.

**Comment by r7kamura** [comment]
I am trying to add such API for 3rd party template plugins to rubocop gem.

Maybe by modifying a little around `RuboCop::Runner#inspect_file`, where file path is converted  to ProcessedSource, it should be possible to generate multiple ProcessedSource from a single file path. And I think it would work if the plugins can change this conversion process according to the file extension.

---

This is the current progress:

- https://github.com/rubocop/rubocop/pull/10839
